### PR TITLE
Do not emit type errors on recovered blocks

### LINKED ISSUE #44579: Weird type error after syntax error
## Steps to Reproduce

Given this code:

```rust
use std::env;

pub struct Foo {
    text: String
}

pub fn parse() -> Foo {
    let args: Vec<String> = env::args().collect();
    let text = args[1].clone();
    
    pub Foo { text }    // NB: Syntax error here
}

fn main() { }
```

## Expected Output

The syntax error is diagnosed

## Actual Output

- There are 2 errors, one actual and one nonsensical
- The nonsensical error uses more visual space (we can't really fault it for that) and it comes last, so it takes all the attention

```rust
error: unmatched visibility `pub`
  --> src/main.rs:11:5
   |
11 |     pub Foo { text }
   |     ^^^

error[E0308]: mismatched types
  --> src/main.rs:7:23
   |
7  |   pub fn parse() -> Foo {
   |  _______________________^
8  | |     let args: Vec<String> = env::args().collect();
9  | |     let text = args[1].clone();
10 | |     
11 | |     pub Foo { text }
12 | | }
   | |_^ expected struct `Foo`, found ()
   |
   = note: expected type `Foo`
              found type `()`
```

I'm sorry to be so harsh on the diagnostics; but we can be better at showing the user relevant diagnostics. (I remember g++ and having to learn to ignore everything after the first error. And those error outputs were long..)

This bug was reported after a discussion in #rust-beginners after a user was struggling with similar code and they came asking about error no. 2. Not about both errors, but about error no. 2.

### PR DESCRIPTION
**Pull request by estebank**
When a parse error occurs on a block, the parser will recover and create
a block with the statements collected until that point. Now a flag
stating that a recovery has been performed in this block is propagated
so that the type checker knows that the type of the block (which will be
identified as `()`) shouldn't be checked against the expectation to
reduce the amount of irrelevant diagnostic errors shown to the user.

Fix #44579.

### CODE REVIEW

**Comment by petrochenkov** [comment]
The case reported in https://github.com/rust-lang/rust/issues/44579 is 100% reasonable recovery.
The "mismatched types" error is exactly what would be reported if the `pub struct Pub` items wasn't ill-formed, so I'd close this as not-a-defect.

(Also, the "unmatched visibility" wording was really weird and unhelpful (unmatched with what?), it's really nice that it's changed to "missing `struct`" on nightly.)

**Comment by petrochenkov** [comment]
Output on nightly, the first error is much more visible now:
```
error: missing `struct` for struct definition
  --> src/main.rs:11:8
   |
11 |     pub Foo { text }    // NB: Syntax error here
   |        ^
help: add `struct` here to parse `Foo` as a public struct
   |
11 |     pub struct Foo { text }    // NB: Syntax error here
   |         ^^^^^^

error[E0308]: mismatched types
  --> src/main.rs:7:23
   |
7  |   pub fn parse() -> Foo {
   |  _______________________^
8  | |     let args: Vec<String> = env::args().collect();
9  | |     let text = args[1].clone();
10 | |     
11 | |     pub Foo { text }    // NB: Syntax error here
12 | | }
   | |_^ expected struct `Foo`, found ()
   |
   = note: expected type `Foo`
              found type `()`
```

**Comment by estebank** [comment]
@petrochenkov the first error is more prominent now _in this case_, but any syntax error will cause the type error to happen. Consider the following case instead:

```rust
pub struct Foo {
    text: String
}

pub fn parse() -> Foo {
    fn
    Foo { text: "".to_string() }
}
```

```
error: expected one of `(` or `<`, found `{`
 --> src/main.rs:7:9
  |
7 |     Foo { text: "".to_string() }
  |         ^ expected one of `(` or `<` here

error[E0308]: mismatched types
 --> src/main.rs:5:23
  |
5 |   pub fn parse() -> Foo {
  |  _______________________^
6 | |     fn
7 | |     Foo { text: "".to_string() }
8 | | }
  | |_^ expected struct `Foo`, found ()
  |
  = note: expected type `Foo`
             found type `()`

error: aborting due to 2 previous errors
```

The second error won't happen once the syntax error is fixed, but the compiler can't know that, and it is still the most vertically prominent error in the output. I feel that masking the type error when we know the block had a parse error (that we made an effort to continue after) is a better default.

**Comment by petrochenkov** [comment]
If some piece of code look so similar to a legal statement that we recover it as a legal statement (e.g. an item statement in case of `pub struct Foo`), then it should not poison the outer block in which it's defined, only possibly its inner block.
```rust
pub fn parse() -> Foo { <not_poisoned>
    let args: Vec<String> = env::args().collect();
    let text = args[1].clone();
    
    pub Foo { <poisoned> text </poisoned> }
</not_poisoned> }
```

In the example with `fn` it looks like we don't recover `fn Foo { text: "".to_string() }` as a function item statement, it's interpreted as just some random token soup. In this case it should indeed poison the block in which it's defined.
```rust
pub fn parse() -> Foo { <poisoned>
    fn
    Foo { text: "".to_string() }
</poisoned> }
```

I need to make some experiments.

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4311,12 +4313,13 @@ impl<'a> Parser<'a> {
...
             rules: s,
             span: lo.to(self.prev_span),
+            recovered,
         }))
     }
 
     /// Parse a statement, including the trailing semicolon.
     pub fn parse_full_stmt(&mut self, macro_legacy_warnings: bool) -> PResult<'a, Option<Stmt>> {
-        let mut stmt = match self.parse_stmt_(macro_legacy_warnings) {
+        let mut stmt = match self.parse_stmt_without_recovery(macro_legacy_warnings)? {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: This seems to be a change independent from adding the `recovered` flag to blocks?
From what I see in the test diffs I'd keep this recovery in place.

File: src/test/ui/mismatched_types/recovered-block.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    text: String
+}
+
+pub fn parse() -> Foo {
+    let args: Vec<String> = env::args().collect();
+    let text = args[1].clone();
+
+    pub Foo { text }
+}
+//~^^ ERROR missing `struct` for struct definition
```
Comment (petrochenkov) [inline src/test/ui/mismatched_types/recovered-block.rs]: Could you add the example with `fn` (https://github.com/rust-lang/rust/pull/46732#issuecomment-351846430) as a test too?

**Comment by petrochenkov** [comment]
r=me with comments addressed

I still want to look at the statement recovery more carefully, but not right now (probably replace the `recovered` flag with `ExprKind::Err` and statements based on it). Into backlog it goes...

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: The recovery happens in the callee already, and using `parse_stmt_` swallows errors that we do want to see in `parse_block_tail`.

File: src/test/ui/mismatched_types/recovered-block.rs (hunk 2 above)
Comment (estebank) [reply src/test/ui/mismatched_types/recovered-block.rs]: added

**Comment by estebank** [comment]
@bors r-

`rustfmt` needs to be updated with this change for this PR to be able to land:

```
error[E0063]: missing field `recovered` in initializer of `syntax::ast::Block`
   --> src\tools\rustfmt\src\closures.rs:127:17
    |
127 |     let block = ast::Block {
    |                 ^^^^^^^^^^ missing `recovered`
error: aborting due to previous error
```

How do we deal with cross-repo commits?

**Comment by petrochenkov** [comment]
>How do we deal with cross-repo commits?

rustfmt and rls are marked as "Broken" in https://github.com/rust-lang/rust/blob/master/src/tools/toolstate.toml, a pull request to rustfmt can be sent asyncroniously.

**Comment by estebank** [comment]
Already marked as broken, looks like the failure was due to sscache.

@bors r=petrochenkov

**Comment by kennytm** [comment]
RLS and rustfmt are still broken.

**Pull request merged**